File: pset6/credit/credit.py
import re

while True:
    num = input("Number: ")
    x = len(num)
    try: 
        a = int(num)
    except ValueError:  # reject ones that dont work
        continue
    break
    

def main():
    if checksum(num):
        if re.search("^4", num):
            if x == 13 or x == 16:  # is it visa
                print("VISA")
            else:
                print("INVALID")
        elif re.search("^3[47]", num):  # is it amex
            if x == 15:
                print("AMEX")
            else:
                print("INVALID")
        elif re.search("^5[1-5]", num):  # is it mastercard
            if x == 16:
                print("MASTERCARD")
            else:
                print("INVALID")
        else: 
            print("INVALID")  # other case
    else: 
        print("INVALID")


def checksum(num):
    y = 0  # yeah that
    
    for i in range(x - 2, -1, -2):  # the two digit things
        z = 2 * int(num[i])
        y += (z // 10 + z % 10) 
    
    for j in range(x - 1, -1, -2):
        y += int(num[j])
    
    if y % 10 == 0:  # is it valid
        return True
    else:
        return False


main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool checksum(long n);

int main(void)
{
    long num = get_long("Number: ");
    if (checksum(num))
    {
        if (4000000000000 <= num && num < 5000000000000) //check to make sure theyre in the right ranges
        {
            printf("VISA\n");
        }
        
        else if (340000000000000 <= num && num < 350000000000000)
        {
            printf("AMEX\n");
        }
        
        else if (370000000000000 <= num && num < 380000000000000)
        {
            printf("AMEX\n");
        }
        
        else if (4000000000000000 <= num && num < 5000000000000000)
        {
            printf("VISA\n");
        }
        
        else if (5100000000000000 <= num && num < 5600000000000000) //too bad no power function
        {
            printf("MASTERCARD\n"); //ok i dont think there was a problem with the ordering of intervals 
            //all the problems were in the checksum function lol
        }
        
        else 
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

bool checksum(long n)
{
    long x = 0;
    for (long i = 10; i < 10 * n + 1; i *= 100) //has to be long int or else zero division bc it gets too big
    {
        x += (n % i) / (i / 10); //for the odd places
    }
    
    for (long j = 100; j < 10 * n + 1; j *= 100) //for the even places
    {
        //x += 2 * (n % j) / (j / 10);
        x += (2 * ((n % j) / (j / 10))) % 10; //first digit
        x += (2 * ((n % j) / (j / 10))) / 10; //second digit
        /*so if i wanted to do both on one line i should do parentheses over it because += some expression
        then + another would've gone weird*/
    }
    if (x % 10 == 0) //this is straightforward
    {
        return true;
    }
    else
    {
        return false;
    }
}